apps/meteor/client/views/room/composer/messageBox/TextMessageBox.tsx

```diff
@@ -0,0 +1,201 @@
+import type { FormattingButton } from '../../../../../app/ui-message/client/messageBox/messageBoxFormatting';
+import { formattingButtons } from '../../../../../app/ui-message/client/messageBox/messageBoxFormatting';
+import type { ComposerAPI } from '../../../../lib/chats/ChatAPI';
+import { keyCodes } from '../../../../lib/utils/keyCodes';
+import { useComposerPopup } from '../../contexts/ComposerPopupContext';
+import ComposerBoxPopup from '../ComposerBoxPopup';
+import ComposerBoxPopupPreview from '../ComposerBoxPopupPreview';
+import { useAutoGrow } from '../RoomComposer/hooks/useAutoGrow';
+import { useComposerBoxPopup } from '../hooks/useComposerBoxPopup';
+import { useEnablePopupPreview } from '../hooks/useEnablePopupPreview';
+import { useMessageComposerMergedRefs } from '../hooks/useMessageComposerMergedRefs';
+import MessageBoxFormattingToolbar from './MessageBoxFormattingToolbar';
+import { useMessageBoxAutoFocus } from './hooks/useMessageBoxAutoFocus';
+
+const handleFormattingShortcut = (
+	event: KeyboardEvent<HTMLTextAreaElement>,
+	formattingButtons: FormattingButton[],
+	composer: ComposerAPI,
+) => {
+	const isMacOS = navigator.platform.indexOf('Mac') !== -1;
+	const isCmdOrCtrlPressed = (isMacOS && event.metaKey) || (!isMacOS && event.ctrlKey);
+
+	if (!isCmdOrCtrlPressed) {
+		return false;
+	}
+
+	const key = event.key.toLowerCase();
+
+	const formatter = formattingButtons.find((formatter) => 'command' in formatter && formatter.command === key);
+
+	if (!formatter || !('pattern' in formatter)) {
+		return false;
+	}
+
+	composer.wrapSelection(formatter.pattern);
+	return true;
+};
+
+const emptySubscribe = () => () => undefined;
+const a: any[] = [];
+const getEmptyArray = () => a;
+
+type TextMessageBoxProps = {
+	rid: IRoom['_id'];
+	tmid?: IMessage['_id'];
+	onResize?: () => void;
+	onTyping?: () => void;
+	onEscape?: () => void;
+	onChange?: FormEventHandler<HTMLElement>;
+	tshow?: IMessage['tshow'];
+	previewUrls?: string[];
+	subscription?: ISubscription;
+	showFormattingTips: boolean;
+	isEmbedded?: boolean;
+};
+
+const TextMessageBox = ({ onTyping, onChange, rid }: TextMessageBoxProps): ReactElement => {
+	const [composer, setComposer] = useState<ComposerAPI>();
+	const t = useTranslation();
+	const composerPlaceholder = t('Message');
+
+	const { isMobile } = useLayout();
+	const sendOnEnterBehavior = useUserPreference<'normal' | 'alternative' | 'desktop'>('sendOnEnter') || isMobile;
+	const sendOnEnter = sendOnEnterBehavior == null || sendOnEnterBehavior === 'normal' || (sendOnEnterBehavior === 'desktop' && !isMobile);
+
+	const textareaRef = useRef<HTMLTextAreaElement>(null);
+	const messageComposerRef = useRef<HTMLElement>(null);
+	const shadowRef = useRef<HTMLDivElement>(null);
+
+	const storageID = `messagebox_${rid}-fileupload`;
+	const callbackRef = useCallback(
+		(node: HTMLTextAreaElement) => {
+			if (node === null || composer) {
+				return;
+			}
+			setComposer(createComposerAPI(node, storageID));
```

Comment:
The composer is set, but never cleared. Since there's no "send" button in this component, this action must be triggered bu its parent component (the upload modal, in this case). I'm not sure what would be the best way to tackle this :thinking: do we do something similar anywhere else?